jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,517 @@
+     * @param rateControlFactory Factory to create a Rate per Tracker
+     * @param rejectHandler A {@link Handler} used to reject excess requests, or {@code null} for a default.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DosHandler(
+        @Name("getId") Function<Request, String> getId,
+        @Name("rateFactory") RateControlFactory rateControlFactory,
+        @Name("rejectHandler") Request.Handler rejectHandler,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        this(null, getId, rateControlFactory, rejectHandler, maxTrackers);
+    }
+
+    /**
+     * @param handler Then next {@link Handler} or {@code null}
+     * @param getId Function to extract an remote ID from a request.
+     * @param rateControlFactory Factory to create a Rate per Tracker
+     * @param rejectHandler A {@link Handler} used to reject excess requests, or {@code null} for a default.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DosHandler(
+        @Name("handler") Handler handler,
+        @Name("getId") Function<Request, String> getId,
+        @Name("rateFactory") RateControlFactory rateControlFactory,
+        @Name("rejectHandler") Request.Handler rejectHandler,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        super(handler);
+        installBean(_trackers);
+        _getId = Objects.requireNonNullElse(getId, ID_FROM_REMOTE_ADDRESS);
+        installBean(_getId);
+        _rateControlFactory = Objects.requireNonNull(rateControlFactory);
+        installBean(_rateControlFactory);
+        _maxTrackers = maxTrackers <= 0 ? 10_000 : maxTrackers;
+        _rejectHandler = Objects.requireNonNullElseGet(rejectHandler, DelayedEnhanceYourCalmRejectHandler::new);
+        installBean(_rejectHandler);
+    }
+
+    @Override
+    public void setServer(Server server)
+    {
+        super.setServer(server);
+        if (_rejectHandler instanceof Handler handler)
+            handler.setServer(server);
+    }
+
+    @Override
+    protected boolean onConditionsMet(Request request, Response response, Callback callback) throws Exception
+    {
+        // Reject if we have too many Trackers
+        if (_maxTrackers > 0 && _trackers.size() > _maxTrackers)
+            return _rejectHandler.handle(request, response, callback);
+
+        // Calculate an id for the request (which may be global empty string)
+        String id;
+        id = _getId.apply(request);
+
+        if (id == null)
+            return _rejectHandler.handle(request, response, callback);
+
+        // Obtain a tracker
+        Tracker tracker = _trackers.computeIfAbsent(id, this::newTracker);
```

Comment:
I can't see where the `Tracker` is cleared from the map when all requests from this ID are done?

If it is not done, eventually if it keeps getting requests with different IDs we will reach `_maxTrackers` and never be able to serve any new requests.

Reply:
See the `CyclicTimouts` usage.  If the `onExpired` method returns true, then the `Tracker` is removed from the iterator, removing it from the map.